File: frontend/src/api/geospatial.ts
import { apiRequest } from "./client";
import { asResolveResponse } from "./validate";
import type { GeoResolveRequest, GeoResolveResponse } from "./types";

/**
 * Resolve a place name or bounding box into a validated geographic
 * representation (center point + bounding box) via the backend.
 */
export async function resolveLocation(
  request: GeoResolveRequest,
  signal?: AbortSignal,
): Promise<GeoResolveResponse> {
  // Coordinates go straight to the map camera; a missing one would place the
  // viewport at NaN rather than fail.
  return asResolveResponse(
    await apiRequest<unknown>("/api/v1/geospatial/resolve", {
      method: "POST",
      body: request,
      signal,
    }),
  );
}

File: backend/app/api/routes/geospatial.py
"""Geospatial grounding endpoints."""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.limits import rate_limited
from app.services.geospatial import GeospatialService, ResolveRequest, ResolveResponse

router = APIRouter()


def get_geospatial_service() -> GeospatialService:
    """Provider for :class:`GeospatialService`; overridden in tests."""

    return GeospatialService()


@router.post(
    "/resolve",
    response_model=ResolveResponse,
    # Reaches a third party whose usage policy this application must honour.
    # The geocoder's own budget is application-wide; this bounds one caller.
    dependencies=[Depends(rate_limited)],
)
async def resolve_location(
    request: ResolveRequest,
    service: GeospatialService = Depends(get_geospatial_service),
) -> ResolveResponse:
    """Resolve a place name (via OpenStreetMap Nominatim) or a bounding box into
    a validated geographic representation (center point + bounding box)."""

    return await service.resolve(request)
